subsys/bluetooth/services/hrs_client.c

```diff
@@ -0,0 +1,414 @@
+
+	if (write_cb) {
+		write_cb(hrs_c, err);
+	}
+}
+
+int bt_hrs_client_measurement_subscribe(struct bt_hrs_client *hrs_c,
+					bt_hrs_client_notify_cb notify_cb)
+{
+	int err;
+	struct bt_gatt_subscribe_params *params = &hrs_c->measurement_char.notify_params;
+
+	if (!hrs_c || !notify_cb) {
+		return -EINVAL;
+	}
+
+	if (atomic_test_and_set_bit(&hrs_c->state, HRS_MEASUREMENT_NOTIFY_ENABLED)) {
+		LOG_ERR("Heart Rate Measurement characterisic notifications already enabled.");
+		return -EALREADY;
+	}
+
+	hrs_c->measurement_char.notify_cb = notify_cb;
+
+	params->ccc_handle = hrs_c->measurement_char.ccc_handle;
+	params->value_handle = hrs_c->measurement_char.handle;
+	params->value = BT_GATT_CCC_NOTIFY;
+	params->notify = on_hrs_measurement_notify;
+
+	atomic_set_bit(params->flags, BT_GATT_SUBSCRIBE_FLAG_VOLATILE);
+
+	err = bt_gatt_subscribe(hrs_c->conn, params);
+	if (err) {
+		atomic_clear_bit(&hrs_c->state, HRS_MEASUREMENT_NOTIFY_ENABLED);
+		LOG_ERR("Subscribe to Heart Rate Measurement characteristic failed");
+	} else {
+		LOG_DBG("Subscribed to Heart Rate Measurement characteristic");
+	}
+
+	return err;
+}
+
+int bt_hrs_client_measurement_unsubscribe(struct bt_hrs_client *hrs_c)
+{
+	int err;
+
+	if (!hrs_c) {
+		return -EINVAL;
+	}
+
+	if (!atomic_test_bit(&hrs_c->state, HRS_MEASUREMENT_NOTIFY_ENABLED)) {
+		return -EFAULT;
+	}
+
+	err = bt_gatt_unsubscribe(hrs_c->conn, &hrs_c->measurement_char.notify_params);
+	if (err) {
+		LOG_ERR("Unsubscribing from Heart Rate Measurement characteristic failed, err: %d",
+			err);
+		atomic_clear_bit(&hrs_c->state, HRS_MEASUREMENT_NOTIFY_ENABLED);
+	} else {
+		LOG_DBG("Unsubscribed from Heart Rate Measurement characteristic");
+	}
+
+	return err;
+}
+
+int bt_hrs_client_sensor_location_read(struct bt_hrs_client *hrs_c,
+				      bt_hrs_client_read_sensor_location_cb read_cb)
+{
+	int err;
+	struct bt_gatt_read_params *params;
+
+	if (!hrs_c || !read_cb) {
+		return -EINVAL;
+	}
+
+	if (atomic_test_and_set_bit(&hrs_c->state, HRS_SENSOR_LOCATION_READ_IN_PROGRES)) {
+		return -EBUSY;
+	}
+
+	hrs_c->sensor_location_char.read_cb = read_cb;
+	params = &hrs_c->sensor_location_char.read_params;
+
+	params->handle_count = 1;
+	params->single.handle = hrs_c->sensor_location_char.handle;
+	params->single.offset = 0;
+	params->func = on_hrs_sensor_location_read;
+
+	err = bt_gatt_read(hrs_c->conn, params);
+	if (err) {
+		LOG_ERR("Reading Heart Rate Measurement characteristic failed, err: %d", err);
+		atomic_clear_bit(&hrs_c->state, HRS_SENSOR_LOCATION_READ_IN_PROGRES);
+	}
+
+	return err;
+}
+
+bool bt_hrs_client_has_sensor_location(struct bt_hrs_client *hrs_c)
+{
+	return (hrs_c->sensor_location_char.handle != 0);
+}
+
+int bt_hrs_client_control_point_write(struct bt_hrs_client *hrs_c,
+				      enum bt_hrs_client_cp_value value,
+				      bt_hrs_client_write_cb write_cb)
+{
+	int err;
+	struct bt_gatt_write_params *params;
+
+	if (!hrs_c) {
+		return -EINVAL;
+	}
+
```

Comment:
Shouldn't you verify if this characteristic is available here by calling bt_hrs_client_has_control_point?

Reply:
Ok